### ISSUE TITLE
confusion about the batchsize parameter in trainer with batchsize parameter in  sampler 

### ISSUE THREAD

**Issue by yuxiaokang-source** [KevinMusgrave/pytorch-metric-learning#258]
I found it's difficult to understand the batchsize parameter between the sampler and trainer.
for example,for mnist data, train_dataset have 60000 samples,
the sampler and trainer are defined below,
```
sampler = samplers.MPerClassSampler(y_train.flatten(), m=4,batchsize=32, length_before_new_iter=len(train_dataset))
```
```
trainer = trainers.MetricLossOnly(models,
                                  optimizers,
                                  batch_size=64,
                                  loss_funcs,
                                  mining_funcs,
                                  train_dataset,
                                  sampler=sampler,
                                  dataloader_num_workers = 0,
                                  end_of_iteration_hook = hooks.end_of_iteration_hook,
                                  end_of_epoch_hook = end_of_epoch_hook)

```
for only a batch training process, how many samples have I choosed to, 4*10 or 64?
for a epoch training process, how many batchs will I have  in a epoch(60000/64 or 60000/40)?Could you give me some explanations ? thank you for you help! 

After I debug in the function,only when
```
 batchsize(sampler)==m*num_of_class and batchsize(trainer) % batchsize(sampler)==0
```
I can get the balanced batch data in every batch training process.

**Comment by KevinMusgrave** [comment]
Good questions

## trainer.batch_size and MPerClassSampler.batch_size

Setting ```batch_size = 64``` for the trainer will cause the dataloader to return 64 samples at every iteration. So the number of batches in an epoch will be 60000/64.

To guarantee that every batch contains exactly 4 samples per class, you should set the MPerClassSampler ```batch_size``` to the same value as the trainer ```batch_size```. So in your example, you should set ```batch_size = 64```. 

If you set the trainer batch size to 64 and the sampler batch size to 32, then each batch will still have size 64. The sampler batch size does not determine the batch size returned by the dataloader. Instead, the ```batch_size``` argument is used inside of MPerClassSampler to guarantee that every batch contains ```m``` samples per class.


## Allowable values for ```m``` and ```batch_size```
MNIST has 10 labels, so if you pass in ```m=4``` and ```batch_size=64``` into MPerClassSampler, you should get an error. This is because a batch of 64 with 4 samples per class must have 64/4 = 16 labels. But there are only 10 labels in MNIST.

So instead, try a value like ```m=8```. Then each batch will contain 8 samples per class, and there will be 64/8 = 8 classes in each batch.

**Comment by KennyThinh** [comment]
Am I right to understand that m in MPerClassSampler is "the smallest number of samples per class needed on each batch_size of trainer"?
For example, if there are 10 classes, 
- if I set m=4, and batch_size (of trainer) = 64 --> there are 64 images per batch. each class has 4 images --> 40 images. How about the remaining 24? They are randomly choosing among these 10 classes?
- if I set m=7, and batch_size (of trainer) = 64 --> will it be error because they need at least m*n_classes = 70 images?

**Comment by KevinMusgrave** [comment]
> For example, if there are 10 classes,
> 
> * if I set m=4, and batch_size (of trainer) = 64 --> there are 64 images per batch. each class has 4 images --> 40 images. How about the remaining 24? They are randomly choosing among these 10 classes?

There will be (4 images of 10 classes = 40) + (4 images of 6 classes = 24). So in the entire batch, there will be 4 images of some classes and 8 images of others.

If you pass ```batch_size``` into the constructor of MPerClassSampler, then it will raise an exception 4*10 is less than 64. 

> * if I set m=7, and batch_size (of trainer) = 64 --> will it be error because they need at least m*n_classes = 70 images?

It will not throw an error unless you explicitly pass ```batch_size``` into the constructor of MPerClassSampler.

**Comment by KennyThinh** [comment]
So, basically, just need to remember that `batch_size (of sampler) / m <= num_classes`
I got it now. Thank you very much.

**Comment by KennyTC** [comment]
Just want to confirm.

Suppose there are 100 classes. 
- If I set m=1 in MPerClassSampler , and batch_size (of trainer) = 64, then samplers will choose 64 random samples but makes sure 1 samples /1 class ? i.e, 64 classes are choosen, skip other 36 classes?
- If I set m=2 in MPerClassSampler , and batch_size (of trainer) = 64, then samplers will choose 64 random samples but makes sure 2 samples/1 class ? i.e, just 32 classes are chosen, and skip other 68 classes?
